clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -539,20 +539,22 @@ public boolean isUpdateDue(long now) {
         public long maybeUpdate(long now) {
             // should we update our metadata?
             long timeToNextMetadataUpdate = metadata.timeToNextUpdate(now);
-            long timeToNextReconnectAttempt = Math.max(this.lastNoNodeAvailableMs + metadata.refreshBackoff() - now, 0);
```

Comment:
Correct. If all nodes are unavailable because of them are all blacked out, we may wait for shortest remaining reconnect backoff by checking `connectionStates.isBlackedOut`. Otherwise, if `leastLoadedNode` returned null because some nodes are exhausted it's in-flight count, then it should make sense to retry immediately on next round since each `client.poll()` has a chance to free some of in-flight requests by request completion or timeout.

> By the way, seems like we can now remove Metadata.refreshBackoff()?

Good catch! Removing.